Add Versioning Documentation (#1388)
open-telemetry/opentelemetry-go: VERSIONING.md

      must be included as a `/vN` at the end of the module paths used in
      `go.mod` files (e.g., `module go.opentelemetry.io/otel/v2`, `require
      go.opentelemetry.io/otel/v2 v2.0.1`) and in the package import path
      (e.g., `import "go.opentelemetry.io/otel/v2/trace"`). This includes the
      paths used in `go get` commands (e.g., `go get
      go.opentelemetry.io/otel/v2@v2.0.1`.  Note there is both a `/v2` and a
      `@v2.0.1` in that example. One way to think about it is that the module
      name now includes the `/v2`, so include `/v2` whenever you are using the
      module name).
    * If a module is version `v0` or `v1`, do not include the major version in
      either the module path or the import path.
  * Modules will be used to encapsulate signals and components.
    * Experimental modules still under active development will be versioned at
      `v0` to imply the stability guarantee defined by
      [semver](https://semver.org/spec/v2.0.0.html#spec-item-4).

      > Major version zero (0.y.z) is for initial development. Anything MAY
      > change at any time. The public API SHOULD NOT be considered stable.

>     * Mature modules that we guarantee a stable public API will be versioned

Reviewer: ```suggestion
    * Mature modules for which we guarantee a stable public API will be versioned
```

  modules](https://github.com/golang/go/wiki/Modules).
  * [Semantic import
    versioning](https://github.com/golang/go/wiki/Modules#semantic-import-versioning)
    will be used.
    * Versions will comply with [semver 2.0](https://semver.org/spec/v2.0.0.html).
    * If a module is version `v2` or higher, the
      major version of the module must be included as a `/vN` at the end of the
      module paths used in `go.mod` files (e.g., `module
      go.opentelemetry.io/contrib/instrumentation/host/v2`, `require
      go.opentelemetry.io/contrib/instrumentation/host/v2 v2.0.1`) and in the
      package import path (e.g., `import
      "go.opentelemetry.io/contrib/instrumentation/host/v2"`). This includes
      the paths used in `go get` commands (e.g., `go get
      go.opentelemetry.io/contrib/instrumentation/host/v2@v2.0.1`.  Note there
      is both a `/v2` and a `@v2.0.1` in that example. One way to think about
      it is that the module name now includes the `/v2`, so include `/v2`
      whenever you are using the module name).
    * If a module is version `v0` or `v1`, do not include the major version
      in either the module path or the import path.

>   * In addition to public APIs, telemetry produced by stable instrumentation

Reviewer 2: Do we need to define stable instrumentation?  E.g. can new labels be added?  Can the values of a label change?

Author: Yeah, I can take another try and nailing down what stable telemetry means for stable instrumentation. I'm open to suggestions if you have any. I'm thinking the telemetry can accrete but not reduce or change. So new labels can be added, but label values or keys cannot change.

Reviewer 2: It depends on what we are trying to accomplish.  [Kubernetes, for example](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md#stability-classes), cares primarily about ingestion backwards-compatibility, rather than dashboard/alert backwards-compatibility.  Some metric backends are table-based, so adding (or removing) label keys isn't considered backwards-compatible for all metric backends. However, kubernetes doesn't require alert/dashboard stability, so label values can be added or removed, and metric values can be changed (e.g. by adding new streams).

I was involved with the kubernetes definition, so i'm biased :).

Label value stability is hard to define if the underlying thing it is measuring is allowed to change.  E.g. if an experimental processor is renamed, a metric about all processors with a processor_name label key would have a value that changes.

Reviewer: @dashpole this is a great question!  I believe one of the goals of OpenCensus was to make it possible to add labels without breaking compatibility, to allow instrumentation to evolve and/or for users to configure which label dimensions they use. A frequent question we get is about the distinction between `UpDownSumObserver` and `ValueObserver`. 

One way to explain this distinction is that it supports automatic label removal without changing semantics. When removing labels from an UpDownSumObserver we sum the corresponding data points into a timeseries of the sum. When removing labels from a ValueObserver, we simply remove labels into a timeseries of reduced-dimension values.

The OTel-Go Metrics SDK has a "reducer processor" that implements label removal, for example, which is a mechanism OTel has to support OpenCensus views. Handling this "semantic removal of labels" in the OTel Collector requires more than just stripping labels, and is probably something the project will need for compatibility with "table-based" systems.

Author: @dashpole, if I understand you correctly, you're saying this part of [the otep]:

> Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard. This means that existing data may not be mutated or removed without a major version bump. Additional data may be added. This applies to spans, metrics, resources, attributes, events, and any other data types that OpenTelemetry emits.

Is unrealizable? Or, that I need to change the wording here to match that better?

Reviewer 2: Unrealizable.  It looks like Josh, my TL, has already raised the same point in the OTEP: https://github.com/open-telemetry/oteps/pull/143#discussion_r541787167, which is probably a better place to have that discussion.  We can just wait for the outcome there and follow.  The wording here matches what is in the OTEP.

Author: With the merging of the linked OTEP we are going to stick with this language and adjust accordingly if the project language changes.

    will remain stable and backwards compatible. This is to avoid breaking
    alerts and dashboard.
  * Modules will be used to encapsulate instrumentation, detectors, exporters,
    propagators, and any other independent sets of related components.
    * Experimental modules still under active development will be versioned at
      `v0` to imply the stability guarantee defined by
      [semver](https://semver.org/spec/v2.0.0.html#spec-item-4).

      > Major version zero (0.y.z) is for initial development. Anything MAY
      > change at any time. The public API SHOULD NOT be considered stable.

>     * Mature modules that we guarantee a stable public API and telemetry will

Reviewer: ```suggestion
    * Mature modules for which we guarantee a stable public API and telemetry will
```

These modules have been developed to a point where the `otel/trace`,
`otel/baggage`, and `otel/sdk/trace` modules have reached a point that they
should be considered for a stable release. The `otel/metrics` and
`otel/sdk/metrics` are still under active development and the `otel` module
depends on both `otel/trace` and `otel/metrics`.

The `otel` package is refactored to remove its dependencies on `otel/metrics`
so it can be released as stable as well. With that done the following release
candidates are made:

* `otel`: `v1.0.0-rc.1`
* `otel/trace`: `v1.0.0-rc.1`
* `otel/baggage`: `v1.0.0-rc.1`
* `otel/sdk/trace`: `v1.0.0-rc.1`

The `otel/metrics` and `otel/sdk/metrics` modules remain at `v0.14.0`.

A few minor issues are discovered in the `otel/trace` package. These issues are
resolved with some minor, but backwards incompatible, changes and are released
as a second release candidate:

* `otel`: `v1.0.0-rc.2`
* `otel/trace`: `v1.0.0-rc.2`
* `otel/baggage`: `v1.0.0-rc.2`
* `otel/sdk/trace`: `v1.0.0-rc.2`

Notice that all module version numbers are incremented to adhere to our
versioning policy.

After these release candidates have been evaluated to satisfaction, they are

> released as version `v1.0.0`.

Reviewer 3: It may be worth mentioning that according to `go mod` sorting rules `v1.0.0` is considered newer than `v1.0.0-rc.2`, which is exactly what we want to happen.

Author: Added after the following codeblock.

* `otel`: `v1.0.0`
* `otel/trace`: `v1.0.0`
* `otel/baggage`: `v1.0.0`
* `otel/sdk/trace`: `v1.0.0`

Active development of this project continues. The `otel/metrics` module now has
backwards incompatible changes to its API that need to be released and the
`otel/baggage` module has a minor bug fix that needs to be released. The
following release is made:

* `otel`: `v1.0.1`
* `otel/trace`: `v1.0.1`

> * `otel/metrics`: `v0.15.0`

Reviewer 2: If an experimental `otel/logging` module is added, would it start at v0.15.0, given the "modules with the same major version move in lockstep" scheme?

Author: I had written the policy with the intent of having the "modules with the same major version move in lockstep" part only applying to stable modules (i.e. v1+). Meaning that a new `otel/logging` module would start it's versioning at `v0.0.0`.

        released.
  * Contrib modules will be kept up to date with this project's releases.
    * Due to the dependency contrib modules will implicitly have on this
      project's modules the release of stable contrib modules to match the
      released version number will be staggered after this project's release.
      There is no explicit time guarantee for how long after this projects
      release the contrib release will be. Effort should be made to keep them
      as close in time as possible.
    * No additional stable release in this project can be made until the
      contrib repository has a matching stable release.
    * No release can be made in the contrib repository after this project's
      stable release except for a stable release of the contrib repository.
* GitHub releases will be made for all releases.
* Go modules will be made available at Go package mirrors.

## Example Versioning Lifecycle

To better understand the implementation of the above policy the following
example is provided. This project is simplified to include only the following
modules and their versions:

* `otel`: `v0.14.0`
* `otel/trace`: `v0.14.0`

> * `otel/metrics`: `v0.14.0`

Reviewer 4: ```suggestion
* `otel/metric`: `v0.14.0`
```

And all other places below where `otel/metrics` is written.

* `otel/baggage`: `v0.14.0`
* `otel/sdk/trace`: `v0.14.0`

> * `otel/sdk/metrics`: `v0.14.0`

Reviewer 4: ```suggestion
* `otel/sdk/metric`: `v0.14.0`
```

And all other places below where `otel/sdk/metrics` is written.

These modules have been developed to a point where the `otel/trace`,
`otel/baggage`, and `otel/sdk/trace` modules have reached a point that they
should be considered for a stable release. The `otel/metrics` and
`otel/sdk/metrics` are still under active development and the `otel` module
depends on both `otel/trace` and `otel/metrics`.

> The `otel` package is refactored to remove its dependencies on `otel/metrics`

Reviewer 4: This point leaves me with a question if this document should also make a policy on how do we decide if a module should become stable or not. If we assume the situation in this example so far (trace stable, metric experimental), should toplevel module become stable (thus moving/removing a functionality depending on metric) or should stay experimental as long as metric is?

The former policy (make stable, drop experimental code) has a risk of adding awkward code if the module has an interface that uses types or even mentions things from both stable and unstable. I haven't come up with a scenario where awkward code might even be breaking compatibility, but maybe this is also something to be considered.

The latter policy (keep experimental as long as any dependency is experimental) might lead to toplevel module or exporter modules to wait a really long time before reaching stability (like we don't even have logs functionality).

Author: > The former policy (make stable, drop experimental code) has a risk of adding awkward code if the module has an interface that uses types or even mentions things from both stable and unstable. I haven't come up with a scenario where awkward code might even be breaking compatibility, but maybe this is also something to be considered.

That is a good point, adding a method to an interface is a breaking change. If there is a conglomerate interface that includes things from an experimental package, when it is broken apart to allow the module to become stable it cannot be added back in as a part of that interface.

I don't think we have instances of this currently, but it is going to be important to keep this in mind going forward.

> The latter policy (keep experimental as long as any dependency is experimental) might lead to toplevel module or exporter modules to wait a really long time before reaching stability (like we don't even have logs functionality).

This approach would not allow compatibility with new development. If we did wait for metrics to become stable before making `otel` stable it would not be able to account for new signals like logs, events, or errors. That would mean that the `otel` package would never go stable or have to fluctuate back and forth.

Reviewer 4: > > The former policy (make stable, drop experimental code) has a risk of adding awkward code if the module has an interface that uses types or even mentions things from both stable and unstable. I haven't come up with a scenario where awkward code might even be breaking compatibility, but maybe this is also something to be considered.
> 
> That is a good point, adding a method to an interface is a breaking change. If there is a conglomerate interface that includes things from an experimental package, when it is broken apart to allow the module to become stable it cannot be added back in as a part of that interface.
> 
> I don't think we have instances of this currently, but it is going to be important to keep this in mind going forward.

We kinda-sorta may have it with `ProtocolDriver` interface in `exporters/otlp` (#1369). I suppose that in this case, we will need to create an extra interface, just like in golang's `database/sql/driver` there is a main interface `Conn` and some extra interfaces that a `Conn` implementation may also implement (like `Execer`, `ExecerContext` or `ConnBeginTx`).

Still, should the document have the wording about a process of declaring a stable module? Even if it's one sentence about it being a case-by-case thing.

Author: > We kinda-sorta may have it with `ProtocolDriver` interface in `exporters/otlp` (#1369). I suppose that in this case, we will need to create an extra interface, just like in golang's `database/sql/driver` there is a main interface `Conn` and some extra interfaces that a `Conn` implementation may also implement (like `Execer`, `ExecerContext` or `ConnBeginTx`).

Yeah, I think you described the development process I had envisioned.

> Still, should the document have the wording about a process of declaring a stable module? Even if it's one sentence about it being a case-by-case thing.

[Added a sentence](https://github.com/open-telemetry/opentelemetry-go/pull/1388/commits/3d481e69fd58aa0786bc2d2a5914c41992a50ab9):

> The decision to make a module stable will be made on a case-by-case bases by the maintainers of this project.

      (e.g., `import "go.opentelemetry.io/otel/v2/trace"`). This includes the
      paths used in `go get` commands (e.g., `go get
      go.opentelemetry.io/otel/v2@v2.0.1`.  Note there is both a `/v2` and a
      `@v2.0.1` in that example. One way to think about it is that the module
      name now includes the `/v2`, so include `/v2` whenever you are using the
      module name).
    * If a module is version `v0` or `v1`, do not include the major version in
      either the module path or the import path.
  * Modules will be used to encapsulate signals and components.
    * Experimental modules still under active development will be versioned at
      `v0` to imply the stability guarantee defined by
      [semver](https://semver.org/spec/v2.0.0.html#spec-item-4).

      > Major version zero (0.y.z) is for initial development. Anything MAY
      > change at any time. The public API SHOULD NOT be considered stable.

    * Mature modules for which we guarantee a stable public API will be versioned
      with a major version greater than `v0`.
      * The decision to make a module stable will be made on a case-by-case

>         bases by the maintainers of this project.

Reviewer 4: ```suggestion
        basis by the maintainers of this project.
```
